### PR TITLE
Value printing refactor (pretty print structs)

### PR DESCRIPTION
**Pull request by bjorn3**
screenshot_2018-06-08 example rs 2

## TODO

* [x] Find a way to read discriminant of a `Value`
* [x] Test union handling
* [x] Handle generic adt's

### CODE REVIEW

**Comment by bjorn3** [comment]
Blocked on rust-lang/rust#51471

File: src/render/locals.rs [hunk 1]
```diff
@@ -89,87 +91,148 @@ pub fn render_locals<'a, 'tcx: 'a>(ecx: &EvalContext<'a, 'tcx>, frame: Option<&F
...
+                Ok(format!("'{}'", chr))
+            } else {
+                Err(EvalErrorKind::AssumptionNotHeld.into())
+            }
+        }
+        TypeVariants::TyUint(_) => {
+            Ok(format!("{0}", bits))
+        }
+        TypeVariants::TyInt(_) => {
+            Ok(format!("{0}", bits as i128))
```
Comment (oli-obk) [inline src/render/locals.rs]: this should not work for `i64` and smaller. You need to sign extend first

File: src/render/locals.rs [hunk 2]
```diff
@@ -89,87 +91,148 @@ pub fn render_locals<'a, 'tcx: 'a>(ecx: &EvalContext<'a, 'tcx>, frame: Option<&F
...
+            //println!("{:?} {:?} {:?}", val, ty, adt_def.variants);
+            let mut pretty = format!("{} {{ ", ecx.tcx.absolute_item_path_str(adt_def.did).replace("<", "&lt;").replace(">", "&gt;"));
+            for (i, adt_field) in adt_def.variants[variant].fields.iter().enumerate() {
+                let field_pretty: EvalResult<String> = do catch {
+                    let ValTy { value: field_val, ty: field_ty } = ecx.read_field(val, None, ::rustc::mir::Field::new(i), ty)?;
+                    pp_value(ecx, field_ty, field_val)?
+                };
+                pretty.push_str(&format!("{}: {}, ", adt_field.ident.as_str(), match field_pretty {
+                    Ok(field_pretty) => field_pretty,
+                    Err(_err) => "<span style='color: red;'>&lt;err&gt;</span>".to_string(),
```
Comment (oli-obk) [inline src/render/locals.rs]: when does this case occur? if the memory is malformed?

Comment (bjorn3) [reply src/render/locals.rs]: Yes, or not yet initialized.

File: src/render/locals.rs (hunk 1 above)
Comment (bjorn3) [reply src/render/locals.rs]: Fixed

**Comment by bjorn3** [comment]
This works for enums now:

screenshot_2018-08-01 example rs

**Comment by oli-obk** [comment]
Sweet, this makes reading the local table so much better. For larger structs we might want to wrap it in a `<details>` html tag and insert some newlines between the fields.

**Comment by bjorn3** [comment]
> For larger structs we might want to wrap it in a `<details>` html tag and insert some newlines between the fields.

Done

screenshot_2018-08-04 example rs

Build is failing due to broken miri.

**Pull request merged**